instance/models/mixins/secret_keys.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2016 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app model mixins - Database
+"""
+
+# Imports #####################################################################
+
+import os
+import hashlib
+import hmac
+from base64 import b64encode, b64decode
+
+from django.db import models
+from django.template import loader
+
+# Functions ###################################################################
+
+
+def generate_secret_key(char_length):
+    """
+    Creates a base64-encoded string of character length `char_length`
+    for use in other functions.
+    """
+    if char_length <= 0 or char_length % 4:
+        # b64encode will always pad its output to four characters, regardless
+        # of the byte count. We'd rather error out than encode too few or too many.
+        raise ValueError('char_length must be a positive multiple of 4')
+    # We need three random bytes per four random b64 characters.
+    random_bytes = os.urandom(int(char_length * 0.75))
+    # Encode to a b64 string...
+    random_string = b64encode(random_bytes)
+    # ...and return!
+    return random_string
+
+
+# Classes #####################################################################
+
+class SecretKeyInstanceMixin(models.Model):
+    """
+    An instance that needs to generate a set of secret keys for different purposes
+    """
+    secret_key = models.CharField(max_length=48, blank=True)
```

Comment:
I think this is a tad confusing... this is not actually the secret key used for the HMAC digests; it's the base64-encoding of the secret key. So if you want to stick with a base64-encoded secret key field, I think you should rename it to `secret_key_b64` to clarify that the value of this field is not itself the secret key. Alternately, skip the base64-encoding and just store the real secret key in a `BinaryField`. Or, as a final alternative, you could just use the base 64 encoded string as the key; even though some bits would be invariant due to the encoding scheme, it should still have [more than enough entropy](http://security.stackexchange.com/questions/95972/what-are-requirements-for-hmac-secret-key)